### DISCUSSION TITLE
Can't understand what is  library instrumentation

### DISCUSSION THREAD

**Discussion by Roiocam** [open-telemetry/opentelemetry-java-instrumentation#6403] [Q&A] [answered]
about this documrent: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation.md

it is clear to me that what is agent instrumentation, basically using Java instrumentation API and bytebuddy build some hack code.

like recompile JDK classess and put some mesuring code.(this is what I want to do)

but more than just hack(recompile) other library, i have my own library also need mesuring code, so i can use library instrumentation and provider interceptor using OTEL API ?  and then my library user could using those API to create they own meter? 

Hope it's clear what I am asking for.

Best Regard.

**Comment by Roiocam** [comment, marked as answer]
After re-reading the contributed markdown, I seem to understand the meaning of library instrumentation.

there is two way that create monitor for specific library:

1. for those library has callback/interceptor api: using Library instrumentation.
2. for another library without those api: use the Agent instrumentation.

For exmaple:  

`org.apache.kafka.clients.producer.KafkaProducer` has method `Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) `. 

the developer could write some mesuring code on callback.

**Reply by Roiocam** [reply]
I am not sure this conclusion is correct because there are still some documents I haven't finished reading.

**Reply by mateuszrzeszutek** [reply]
Hey @Roiocam ,
You're right about the meaning of the "library instrumentation" phrase. In general, if a library we want to instrument exposes some kind of a hook/listener/interceptor interface (e.g. OkHttp `Interceptor`, Reactor `Hooks`, etc), we create a library instrumentation module that can be used as normal Java library -- without the agent. 
Of course, in most cases we'll have a javaagent instrumentation for that library too; usually it only installs the listener/interceptor implementation from the library instrumentation.
For any other library/framework, or JDK, that does not expose a way to supplement its behavior, we'll implement just the javaagent instrumentation.

**Reply by Roiocam** [reply]
Thank you for your response @mateuszrzeszutek . in addition i notice that on [writing-instrumentation.md](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation.md), point out that developer can use gradle plugin `otel.library-instrumentation`. what this plugin use for ? it is available  on maven ?

Thanks for your time!

**Reply by mateuszrzeszutek** [reply]
That's one of the gradle plugins used in this repo: [this one](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/conventions/src/main/kotlin/otel.library-instrumentation.gradle.kts). It's a set of common settings for library instrumentations: dependencies, static code checks, publishing settings, etc. You probably don't need to care about this outside of this repo.

**Reply by Roiocam** [reply]
thanks, I found those yesterday too. but I am not familiar with Kotlin, so I am not sure how it works inside.  

I use the same structure as `apache-Dubbo-2.7` submodule, but JVM complains NoClassDefineError from the library module class, using `maven-shade-plugin` could fix this error, but I don't see the same operation on  `apache-Dubbo-2.7`.

I could read the plugin/repo source and answer it on my own, plz let me know if this way is correct.

**Reply by mateuszrzeszutek** [reply]
Hmm, I guess it might depend on what you're trying to do. You're implementing a library instrumentation (that is: no javaagent), is that correct? (Library instrumentations use much simpler gradle setup than the javaagent ones)
What is the exact error that you're getting? Is this something connected to the OpenTelemetry code?

**Reply by Roiocam** [reply]
Actually, I am implementing both library instrumentation and agent instrumentation.  the project structure is like this:

```
javaagent ->
    src/main/java ->
        InstrumentationModuleImpl
        TypeInstrumentationImpl (inject Instrumenter API stuff, general using MyInstrumenterImpl method `start` and `end`...)
library ->
    src/main/java ->
        MyInstrumenterImpl (Instrumenter implementation)
        MyMetrics (OperationListener implementation)
```

I am not sure why `apache-dubbo-2.7` don't need to explicitly  override `getAdditionalHelperClassNames` on `InstrumentationModule` implemention. I had to do this to make Instrumentation work, and agent jar(otel.extension) need "shade" library class

**Reply by Roiocam** [reply]
> What is the exact error that you're getting? Is this something connected to the OpenTelemetry code?

after override `getAdditionalHelperClassNames()` and boot jar with my javaagent. JVM complaints can not be found `MyMetrics` class. 

there are two questions:

- why do I have to explicitly override `getAdditionalHelperClassNames ` on `InstrumentationModule`, but `apache-dubbo-2.7` not ?
- why do I have to shade library on my javaagent, but `apache-dubbo-2.7` not? 

I'm wondering what the magic is inside. how `apache-dubbo-2.7` make ClassLoader load the helper class.

**Reply by mateuszrzeszutek** [reply]
> * why do I have to shade library on my javaagent, but `apache-dubbo-2.7` not?

The javaagent shading happens in the [javaagent-shadowing](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/conventions/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-shadowing.gradle.kts) gradle plugin from this repo -- the Dubbo instrumentation automatically includes that as a part of `otel.javaagent-instrumentation` plugin. 

> * why do I have to explicitly override `getAdditionalHelperClassNames ` on `InstrumentationModule`, but `apache-dubbo-2.7` not ?

The [muzzle](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/muzzle.md#compile-time-reference-collection) plugin is responsible for collecting the helper classes during the build time, and injecting them into the `InstrumentationModule` bytecode. Again, it is included by default in all javaagent instrumentation in this repo.

We're publishing [gradle plugins](https://plugins.gradle.org/search?term=muzzle) for muzzle; but you probably won't be able to use them in a maven build.

**Reply by Roiocam** [reply]
Thanks for taking the time to answer. 

seem the same as my thought, I am stuck with those Gradle plugins.

I got all answers in this discussion and I  will read all documents and source code to answer my question in the future. let us get over this discussion which keeps your valuable time.

thanks again for your time.

Best Regards.

**Reply by mateuszrzeszutek** [reply]
No problem at all, you're welcome!

**Comment by Roiocam** [comment]
For anyone who has the same problem as me, [HikariCP](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/hikaricp-3.0) is best example  in helping to understand
